chainer/link_hooks/spectral_normalization.py

```diff
@@ -0,0 +1,252 @@
+                 factor=None, weight_name='W', name=None):
+        assert n_power_iteration > 0
+        self.n_power_iteration = n_power_iteration
+        self.eps = eps
+        self.use_gamma = use_gamma
+        self.factor = factor
+        self.weight_name = weight_name
+        self.vector_name = weight_name + '_u'
+        self._initialied = False
+        self.axis = 0
+        if name is not None:
+            self.name = name
+
+    def __enter__(self):
+        raise NotImplementedError(
+            'This hook is not supposed to be used as context manager.')
+
+    def __exit__(self):
+        raise NotImplementedError
+
+    def added(self, link):
+        # Define axis and register ``u`` if the weight is initialized.
+        if not hasattr(link, self.weight_name):
+            raise ValueError(
+                "Weight '{}' does not exist!".format(self.weight_name))
+        if isinstance(link, (L.Deconvolution2D, L.DeconvolutionND)):
+            self.axis = 1
+        if getattr(link, self.weight_name).array is not None:
+            self._prepare_parameters(link)
+
+    def deleted(self, link):
+        # Remove approximate vector ``u`` and parameter ``gamma` if exists.
+        delattr(link, self.vector_name)
+        if self.use_gamma:
+            del link.gamma
+
+    def forward_preprocess(self, cb_args):
+        # This method normalizes target link's weight spectrally
+        # using power iteration method
+        link = cb_args.link
+        input_variable = cb_args.args[0]
+        if not self._initialied:
+            self._prepare_parameters(link, input_variable)
+        weight = getattr(link, self.weight_name)
+        # For link.W or equivalents to be chainer.Parameter
+        # consistently to users, this hook maintains a reference to
+        # the unnormalized weight.
+        self.original_weight = weight
+        # note: `normalized_weight` is ~chainer.Variable
+        normalized_weight = self.normalize_weight(link, weight)
+        setattr(link, self.weight_name, normalized_weight)
+
+    def forward_postprocess(self, cb_args):
+        # Here, the computational graph is already created,
+        # we can reset link.W or equivalents to be Parameter.
+        link = cb_args.link
+        setattr(link, self.weight_name, self.original_weight)
+
+    def _prepare_parameters(self, link, input_variable=None):
+        """Prepare one buffer and one parameter.
+
+        Args:
+            link (:class:`~chainer.Link`): Link to normalize spectrally.
+            input_variable (:class:`~chainer.Variable`):
+                The first minibatch to initialize weight.
+
+        """
+        if getattr(link, self.weight_name).array is None:
+            if input_variable is not None:
+                link._initialize_params(input_variable.shape[1])
+        initialW = getattr(link, self.weight_name)
+        u = link.xp.random.normal(
+            size=(1, initialW.shape[self.axis])).astype(dtype=link.W.dtype)
```

Comment:
Could you try to make `u` 1-dim?

Another favorable shape is `tuple(s if i == self.axis else 1 for s in initialW.shape)`.  (Don't use `F.reshape` but `F.tensordot`.)